src/cellrank/kernels/_base_kernel.py

```diff
@@ -463,6 +463,135 @@ def _reuse_cache(self, expected_params: Dict[str, Any], *, time: Optional[Any] =
+        graph = self.adata.obsp[graph_key]
+        features = self.adata.obsm[rep]
+        empirical_velo = []
+
+        for boundary_id in boundary_ids:
+            row = graph[boundary_id, :].toarray().squeeze()
+            obs_mask = row.astype(bool) & target_obs_mask
+            neighbors = obs_ids[obs_mask]
+            weights = row[obs_mask]
+
+            empirical_velo.append(
+                np.sum(weights.reshape(-1, 1) * (features[neighbors, :] - features[boundary_id, :]), axis=0)
+            )
+
+        empirical_velo = np.array(empirical_velo)
+        obs_mask = np.isnan(empirical_velo).any(axis=1)
+        empirical_velo = empirical_velo[~obs_mask, :]
+
+        return empirical_velo
+
+    def _get_vector_field_estimate(self, rep: str) -> np.ndarray:
+        """Compute estimate of vector field under one step of the transition matrix.
+
+        Parameters
+        ----------
+        rep
+            Key in :attr:`~anndata.AnnData.obsm` to use as data representation.
+
+        Returns
+        -------
+        Vector field estimate based on kernel dynamics.
+        """
+        extrapolated_gex = self.transition_matrix @ self.adata.obsm[rep]
+        return extrapolated_gex - self.adata.obsm[rep]
+
+    def get_cbc(self, source: str, target: str, cluster_key: str, rep: str, graph_key: str = "distances") -> np.ndarray:
+        """Compute cross-boundary correctness score between source and target cluster.
+
+        Parameters
+        ----------
+        source
+            Name of source cluster.
+        target
+            Name of target cluster.
+        cluster_key
+            Key in :attr:`~anndata.AnnData.obs` to obtain cluster annotations.
+        rep
+            Key in :attr:`~anndata.AnnData.obsm` to use as data representation.
+        graph_key
+            Name of graph representation to use from :attr:`~anndata.AnnData.obsp`.
+
+        Returns
+        -------
+        Cross-boundary correctness score for each observation.
+        """
+
+        def _pearsonr(x, y):
+            x_centered = x - np.mean(x, axis=1).reshape(-1, 1)
+            y_centered = y - np.mean(y, axis=1).reshape(-1, 1)
+            return (
+                np.sum(x_centered * y_centered, axis=1)
+                / np.linalg.norm(x_centered, axis=1)
+                / np.linalg.norm(y_centered, axis=1)
+            )
+
+        target_obs_mask = self.adata.obs[cluster_key].isin([target] if isinstance(target, str) else target)
+        boundary_ids = self._get_boundary(source=source, target=target, cluster_key=cluster_key, graph_key=graph_key)
+        empirical_velo = self._get_empirical_velocity_field(
+            boundary_ids=boundary_ids, target_obs_mask=target_obs_mask, rep=rep, graph_key=graph_key
+        )
+        estimated_velo = self._get_vector_field_estimate(rep=rep)[boundary_ids, :]
+
+        cbc = _pearsonr(x=estimated_velo, y=empirical_velo)
+        if hasattr(self, "cbc"):
```

Comment:
Please remove the `hasattr`, `self.cbc` is never defined.

Reply:
Sorry, another artifact from an older version of the code. I removed the entire if clause; I don't think that we need to save the CBC as a class attribute.